src/DurableTask.SqlServer/Scripts/logic.sql

```diff
@@ -1252,3 +1266,138 @@ BEGIN
     VALUES (@SemanticVersion)
 END
 GO
+
+
+CREATE OR ALTER PROCEDURE dt._RewindInstance
+    @InstanceID varchar(100),
+    @Reason varchar(max) = NULL
+AS
+BEGIN
+    DECLARE @TaskHub varchar(50) = dt.CurrentTaskHub()
+
+    -- *** IMPORTANT ***
+    -- To prevent deadlocks, it is important to maintain consistent table access
+    -- order across all stored procedures that execute within a transaction.
+    -- Table order for this sproc: Instances --> (History --> Payloads --> History), NewEvents)  
+
+    DECLARE @existingStatus varchar(30)
+    DECLARE @executionID varchar(50)
+    
+    SELECT TOP 1 @existingStatus = existing.[RuntimeStatus], @executionID = existing.[ExecutionID]
+    FROM Instances existing WITH (HOLDLOCK)
+    WHERE [TaskHub] = @TaskHub AND [InstanceID] = @InstanceID
+
+    -- Instance IDs can be overwritten only if the orchestration is in a terminal state
+    IF @existingStatus NOT IN ('Failed')
+    BEGIN
+        DECLARE @msg nvarchar(4000) = FORMATMESSAGE('Cannot rewing instance with ID ''%s'' because it is not in a ''Failed'' state, but in ''%s'' state.', @InstanceID, @existingStatus);
+        THROW 50001, @msg, 1;
+    END
+    
+    DECLARE @eventsInFailure TABLE (
+        [SequenceNumber] bigint NULL,
+        [EventType] varchar(40) NULL,
+        [TaskID] int NULL,
+        [DataPayloadID] uniqueidentifier NULL)
+    
+    -- List all events related to failures (ie TaskScheduled/TaskFailed and SubOrchestrationInstanceStarted/SubOrchestrationInstanceFailed couples)
+    INSERT INTO @eventsInFailure
+    SELECT h.[SequenceNumber], h.[EventType], h.[TaskID], h.[DataPayloadID]
+    FROM History h
+    WHERE h.[TaskHub] = @TaskHub AND h.[InstanceID] = @InstanceID
+      AND (h.[EventType] IN ('TaskFailed', 'SubOrchestrationInstanceFailed')
+           OR (h.[EventType] IN ('TaskScheduled', 'SubOrchestrationInstanceStarted') AND EXISTS (SELECT 1
+                                                           FROM History f
+                                                           WHERE f.[TaskHub] = @TaskHub AND f.[InstanceID] = @InstanceID  AND f.[TaskID] = h.[TaskID] AND f.[EventType] = CASE WHEN h.[EventType] = 'TaskScheduled' THEN 'TaskFailed' ELSE 'SubOrchestrationInstanceFailed' END)))
+
+    -- Mark all events related to failure as rewound
+    UPDATE Payloads
+    SET [Reason] = CONCAT('Rewound: ', ef.[EventType])
```

Comment:
This seems a little strange. The `[Reason]` column is supposed to hold runtime data, but it looks like you're using it for rewind-specific metadata. I expect it will "just work" because none of the payload entries we're manipulating need a `[Reason]` value, but I may need to think about this a bit more.

Reply:
I agree with you, I found it a little bit strange.
I chose to replicate as closely as possible the behaviour of Azure Table Storage found [here](https://github.com/Azure/durabletask/blob/12e4c5afc2e6463fd26cfb5f994c180685d45d15/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L357-L358), [here](https://github.com/Azure/durabletask/blob/12e4c5afc2e6463fd26cfb5f994c180685d45d15/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L384), and [here](https://github.com/Azure/durabletask/blob/12e4c5afc2e6463fd26cfb5f994c180685d45d15/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L393-L394), which is basically to keep track of the previous `EventType`, once replaced by `GenericEvent`.

Should `[Reason]` field be made part of the event table, or, if its `varchar(max)` nature has performance impacts, maybe split into its own table ?

Please tell me what direction you would like to take once you've had time to think about it!

Reply:
I think I may have missed this, but are these records being changed to GenericEvent and also having their Reason payloads set? I see that we’re doing this for Azure Storage. If so, then I have no problem with what you’re doing here. I haven’t reviewed very closely yet and thought maybe you were changing values for other event types.

Reply:
Yes that's what this code is doing. It updates payloads for the events who happen to already have one, and then add payloads (for the sole purpose of setting the reason) for those who haven't. Updated events are the ones that happen to have been switched to `GenericEvent`